embrio-async/src/lib.rs

```diff
@@ -110,22 +103,48 @@ where
         // can't have moved between creating the pointer and the generator ever
         // using the pointer so it is safe to dereference).
         let this = unsafe { Pin::get_unchecked_mut(self) };
-        if let FutureImplState::Started(g) = &mut this.state {
-            unsafe {
-                this.context = loosen_context_lifetime(cx);
-                match Pin::new_unchecked(g).resume() {
-                    GeneratorState::Yielded(Poll::Pending) => Poll::Pending,
-                    GeneratorState::Complete(x) => Poll::Ready(x),
-                }
-            }
-        } else if let FutureImplState::NotStarted(c) =
-            mem::replace(&mut this.state, FutureImplState::Invalid)
-        {
-            let gen = c(UnsafeContextRef(&mut this.context));
-            this.state = FutureImplState::Started(gen);
-            unsafe { Pin::new_unchecked(this) }.poll(cx)
-        } else {
-            panic!("reached invalid state")
+        unsafe {
+            this.context.as_mut_ptr().write(loosen_context_lifetime(cx));
+        }
+        match &mut this.state {
+            FutureImplState::Started(g) => unsafe {
+                Pin::new_unchecked(&mut *g.as_mut_ptr())
+            },
+            FutureImplState::NotStarted(..) => unsafe {
+                let c = mem::replace(
+                    &mut this.state,
+                    FutureImplState::Started(MaybeUninit::uninit()),
+                );
+                let c = match c {
+                    FutureImplState::NotStarted(c) => c,
+                    _ => unreachable_unchecked(),
+                };
+                let g = match &mut this.state {
+                    FutureImplState::Started(g) => g,
+                    _ => unreachable_unchecked(),
+                };
```

Comment:
I'm not a huge fan of these matches but they seemed better than the equivalent `if let else unreachable` chains. Feel free to suggest a more succinct approach, or I could pull them out into smaller `.as_started_unchecked()` methods if that makes it a bit more readable.

Reply:
This block might be better if you make the `NotStarted` state contain the `MaybeUninit`, you might be able to replace this with something like

```rust
match &mut this.state {
    FutureImplState::Started(g) => {
        unsafe { this.context.as_mut_ptr().write(loosen_context_lifetime(cx)) };
        unsafe { Pin::new_unchecked(g) }
    }
    FutureImplState::NotStarted(c) => {
        let c = unsafe { c.read() };
        let gen = c(UnsafeContextRef(this.context.as_mut_ptr()));
        *this.state = FutureImplState::Started(gen);
        unsafe { Pin::new_unchecked(this) }.state(cx);
    }
}
```

Then having a `Drop` impl for `FutureImpl` that will drop the closure from `FutureImplState::NotStarted` you have the same "panic while running `C` is bad" safety implication.

Reply:
As mentioned in the other comment, I had `Started` contain the uninitialized field mostly because in my later changes I made a non-stateful `Started` struct that had the `context` field moved into it, which necessitated replacing the enum first before calling the constructor. That's not necessary here and doing it the other way does simplify things, so I'll do that!